### PR TITLE
Fix GitLab OAuth token refresh

### PR DESCRIPTION
**Pull request by chasecaleb**
Fix for #739. This turned out to be tricky to debug but the solution is trivial, fortunately. Here's what was happening, brace yourself:

1. Signed-in user with an expired OAuth token visits https://organice.200ok.ch/files . Tangential-but-irrelevant trivia: GitLab tokens expire after 2 hours and are not refreshed until expiration, so expiration could happen even while actively using the app.
2. On startup, Organice checks if the sync client is still signed in. **Important detail**: because this occurs in a promise that isn't waited on before doing anything else, additional API calls (e.g. to fetch the config file and fetch org file contents) also happen concurrently before this finishes. https://github.com/200ok-ch/organice/blob/891716bcf9c199356a024dfde35e8a9796bdbe42/src/App.js#L168-L174 
3. The GitLab sync client's `isSignedIn()` will refresh the OAuth access token if it is expired by calling `OAuth2AuthCodePKCE.getAccessToken()`. If that it fails, `isSignedIn()` returns false. Keep in mind that a rejected Promise translates to a thrown exception in an `async` function. https://github.com/200ok-ch/organice/blob/891716bcf9c199356a024dfde35e8a9796bdbe42/src/sync_backend_clients/gitlab_sync_backend_client.js#L118-L130
4. Like I pointed out in step 2, there are additional GitLab API calls potentially happening concurrently to `isSignedIn()`. All of the GitLab API calls use `OAuth2AuthCodePKCE.decorateFetchHTTPClient(..)`, which handles OAuth things **including** calling the previously mentioned `getAccessToken()`.
5. As a result, there are multiple concurrent calls to GitLab's token refresh endpoint: https://gitlab.com/oauth/token. All of these refresh calls use identical parameters, which OAuth (at least with 2.0 PKCE) does not allow. All but one of these calls fail with a 400 status and `invalid_grant` error. Side note: according to the OAuth 2.0 spec, `invalid_grant` is pretty much a catch all for "you did something wrong, good luck figuring it out" :facepalm:
6. When `isSignedIn()` returns false, the `App.js` startup logic triggers a sign out action (refer back to step 2 for the code snippet).
7. There's some other problematic behavior going on at this point, but I'm not exactly sure what. Somehow it's possible for `localStorage.authenticatedSyncService` to be cleared without necessarily everything else in `localStorage` being cleared. This is the behavior that we saw here: https://github.com/200ok-ch/organice/issues/739#issuecomment-958107531

**TLDR summary**:
The GitLab backend was fine, turns out the root cause is due to the OAuth library not handling concurrent requests very well if the access token is expired. Fortunately as you can see in this PR, the workaround is simple and isolated to a single location.

---
**Reproduction** from any device, not just mobile:
1. Open dev tools
2. Run this in console:
    ```javascript
          let authstate = JSON.parse(localStorage["oauth2authcodepkce-state"]);
          authstate.accessToken.expiry = new Date("1/1/2020").toString();
          localStorage["oauth2authcodepkce-state"] = JSON.stringify(authstate);
    ````
3. Switch to network requests view of dev tools
4. Ensure that network requests are persisted across page reloads (there's a setting for this somewhere in dev tools, specifics depend on your browser)
5. Reload page
6. Observe multiple calls to https://gitlab.com/oauth/token. One will succeed, the rest will fail with status code 400.

If you do the same with my fix in this PR, instead of multiple token calls in step 6 you will only see a single one.

---

**Tangent**: If the opportunity arises, I think it would be beneficial to improve how side effects are handled in this application, particularly when it comes to asynchronous things like API calls. The craziness that is the constructor of `App` in `src/App.js` is a good example of what I'm referring to.

For example, I've used [Redux Saga](https://redux-saga.js.org/) extensively in a good sized project at work and it worked great. However it can be confusing for people to learn because it's based around the [Command pattern](https://en.wikipedia.org/wiki/Command_pattern) using [JavaScript generator functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/function*), so I'm not necessarily suggesting it as a good fit for this project. There are other libraries for this use case too, e.g. [Redux Thunk](https://github.com/reduxjs/redux-thunk) is another popular option off the top of my head. I'm sure there are quite a few other ways to improve things too.

**However**,  I'm just pointing this out as food for thought someday, I am **not** suggesting any particular changes or solutions since it would be a significant amount of work to deal with. I haven't put much thought into specific solutions or a way to get there without rewriting the world... just pointing out what seems like a pain point from my limited experience with the project so far.

---

I also created an issue for the library as you can see from my comment: https://github.com/BitySA/oauth2-auth-code-pkce/issues/29 . Hopefully we can remove this workaround one day in the future; in the meantime, at least it's a straightforward and well-contained solution.

### CODE REVIEW

**Comment by branch14** [comment]
This is great work and I highly appreciate the effort you put into documenting the issue and your workaround.

LGTM.
